### PR TITLE
Refactor: TokenFile with DerivedState

### PR DESCRIPTION
**Pull request by y-myajima**
# Description
Refactor Token File feature with DerivedState.
Related Issue: https://github.com/AthenZ/k8s-athenz-sia/issues/113

## Assignees
- [x] `Assignees` is set

## Type of changes
- [x] Apply one or more `labels` of the following that fits:
  - `bug`: Bug fix
  - `dependencies`: Dependency upgrades
  - `documentation`: Documentation changes
  - `enhancement`: New Feature
  - `good first issue`: First contribution
  - `logging`: Log changes
  - `refactor`: Refactoring (no functional changes, no api changes)

### Flags
```
- [ ] Breaks backward compatibility
- [ ] Requires a documentation update
- [ ] Has untestable code
```
---

## Checklist
```
- [ ] Followed the guidelines in the CONTRIBUTING document
- [ ] Added prefix `[skip ci]`/`[ci skip]`/`[no ci]`/`[skip actions]`/`[actions skip]` in the PR title if necessary
- [ ] Tested and linted the code
- [ ] Commented the code
- [ ] Made corresponding changes to the documentation
```

## Checklist for maintainer
```
- [ ] Use `Squash and merge`
- [ ] Double-confirm the merge message has prefix `[skip ci]`/`[ci skip]`/`[no ci]`/`[skip actions]`/`[actions skip]`
- [ ] Delete the branch after merge
```

### CODE REVIEW

File: pkg/config/derived-token-file.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+		RoleToken: TokenFileConfig{
+			Use: false,
+			Dir: "",
+			// Format:    "",
+			Delimiter: "",
+		},
+	}
+
+	if idCfg.tokenDir == "" {
+		return nil // disable
```
Comment (mlajkim) [inline pkg/config/derived-token-file.go:54]:
```suggestion
		return nil // disabled
```

Comment (mlajkim) [inline pkg/config/derived-token-file.go:54]:
```suggestion
		return nil // disabled
```

File: pkg/token/service.go [hunk 2]
```diff
@@ -66,9 +66,11 @@ func New(ctx context.Context, idCfg *config.IdentityConfig) (daemon.Daemon, erro
 		return nil, nil
 	}
 	// TODO: In the next PR, the determination will be made on a per Access Token and Role Token basis.
-	enableWriteFiles := idCfg.TokenDir != ""
+	// TODO: move to derived token file
+	enableWriteFiles := idCfg.TokenCache.AccessToken.Use || idCfg.TokenCache.RoleToken.Use
 	if !enableWriteFiles {
```
Comment (mlajkim) [inline pkg/token/service.go:71]: No longer needed
```suggestion
	if !idCfg.TokenCache.AccessToken.Use && !idCfg.TokenCache.RoleToken.Use {
```

File: pkg/config/dervied.go [hunk 3]
```diff
@@ -42,14 +42,18 @@ func (idCfg *IdentityConfig) loadDerivedConfig() error {
 	// TODO:
 	// depends on the following:
 	// - derivedTargetDomainRoles()
-	// if err := idCfg.derivedAccessTokenConfig(); err != nil {
+	// if err := idCfg.derivedTokenCacheConfig(); err != nil {
 	// 	return err
 	// }
 
+	// depends on the following:
```
Comment (mlajkim) [inline pkg/config/dervied.go:49]: Maybe?
```suggestion
```

File: pkg/config/derived-target-domain-roles.go [hunk 4]
```diff
@@ -63,10 +63,9 @@ func (idCfg *IdentityConfig) derivedTargetDomainRoles() error {
 		})
 	}
 
-	idCfg.TokenTargetDomainRoles = tokenDomainRoles // TODO: Delete me and refactor by using the type DerivedTargetDomainRoles below:
```
Comment (mlajkim) [inline pkg/config/derived-target-domain-roles.go:66]: Why is this deleted again?

File: pkg/config/default.go [hunk 5]
```diff
@@ -97,6 +97,7 @@ func DefaultIdentityConfig() *IdentityConfig {
 		roleCertNamingFormat:      "",
 		roleCertKeyNamingFormat:   "",
 		RoleAuthHeader:            DEFAULT_ROLE_AUTH_HEADER,
+		tokenDir:                  "",
```
Comment (mlajkim) [inline pkg/config/default.go:100]: Dont modify the position yet:
https://github.com/AthenZ/k8s-athenz-sia/issues/113#issuecomment-2351386787

File: pkg/config/derived-token-file.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+		},
+		RoleToken: TokenFileConfig{
+			Use: false,
+			Dir: "",
+			// Format:    "",
+			Delimiter: "",
+		},
+	}
+
+	if idCfg.tokenDir == "" {
```
Comment (mlajkim) [inline pkg/config/derived-token-file.go:53]: Wonder if we should have the following:
(not in this PR)
```suggestion
	// TODO: Apply the following instead?:
	// if idCfg.TokenDir == ""  || idCfg.TokenType == "" {
	if idCfg.tokenDir == "" {
```

File: pkg/config/derived-token-file.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+type DerivedTokenFile struct {
+	AccessToken TokenFileConfig
+	RoleToken   TokenFileConfig
+}
+
+// derivedTokenFileConfig reads given configuration and sets the derived state of outputting token file related configuration.
+func (idCfg *IdentityConfig) derivedTokenFileConfig() error {
+	// default:
+	idCfg.TokenCache = DerivedTokenFile{
```
Comment (mlajkim) [inline pkg/config/derived-token-file.go:38]: Where is `Dir`?

File: pkg/config/derived-token-file.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+	// role token:
+	if strings.Contains(idCfg.TokenType, "roletoken") {
+		idCfg.TokenCache.RoleToken = TokenFileConfig{
+			Use: true,
+			Dir: idCfg.tokenDir,
+			// Format:    filepath.Join(idCfg.TokenDir, "{{domain}}{{delimiter}}{{role}}.roletoken"),
+			Delimiter: ":role.",
+		}
+	}
```
Comment (mlajkim) [inline pkg/config/derived-token-file.go:76]: Two:
- Where is Dir?
- Write in anonymous function
```suggestion
        idCfg.TokenCache = ... {
          Dir: "dir"
           ...
        }
```

File: pkg/config/derived-token-file.go [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+type TokenFileConfig struct {
+	Use bool
+	Dir string
+	// TODO: Add Format
+	// Format    string
+	Delimiter string
+}
+
+type DerivedTokenFile struct {
```
Comment (mlajkim) [inline pkg/config/derived-token-file.go:30]: Should have a dir here?

File: pkg/config/derived-token-file.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+// Package config defines all the configuration parameters. It reads configuration from environment variables and command-line arguments.
+package config
+
+import (
+	"strings"
+)
+
+type TokenFileConfig struct {
+	Use bool
+	Dir string
```
Comment (mlajkim) [inline pkg/config/derived-token-file.go:24]: remove me
```suggestion
```

File: pkg/token/service.go [hunk 11]
```diff
@@ -66,9 +66,11 @@ func New(ctx context.Context, idCfg *config.IdentityConfig) (daemon.Daemon, erro
...
-		log.Debugf("Skipping to write token files to directory with empty TOKEN_DIR [%s]", idCfg.TokenDir)
+		// When file output is disabled, the Dir settings for the access token and role token will all be empty strings.
+		log.Debugf("Skipping to write token files to directory with empty TOKEN_DIR [%s]", idCfg.TokenCache.AccessToken.Dir)
```
Comment (mlajkim) [inline pkg/token/service.go:73]:
```suggestion
		log.Debugf("Skipping to write token files to directory with empty TOKEN_DIR [%s]", idCfg.TokenFile.Dir)
```

File: pkg/config/model.go [hunk 12]
```diff
@@ -59,6 +59,10 @@ type IdentityConfig struct {
 	roleCertNamingFormat      string
 	roleCertKeyNamingFormat   string
 	//
+	// Token Cache Derived State and its related fields:
+	TokenCache DerivedTokenFile
```
Comment (mlajkim) [inline pkg/config/model.go:63]:
```suggestion
	TokenFile DerivedTokenFile
```

**Review by mlajkim — COMMENTED** [review]
review

**Review by mlajkim — APPROVED** [review]
ok

**Review by fum1h1to — APPROVED** [review]
